ReadMe for CSDK Release Candidate (#1231)
aws/aws-iot-device-sdk-embedded-C: README.md

> # AWS IoT Device SDK

Reviewer: Why is it just AWS IoT Device SDK, not 'AWS IoT Device SDK for Embedded C'?

  
If you are upgrading from v3.x of the C-SDK to the 202009.00 release, please follow the update guide (LINK REQUIRED). If you are using the C-SDK v4_beta_deprecated branch, note that we will continue to fix bugs but will not add new features to it. See the C-SDK v4_beta_deprecated branch [README](https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/v4_beta_deprecated/README.md) for additional details. 

## Overview

The AWS IoT Device SDK for Embedded C (C-SDK) is a collection of C source files under the [MIT open source license](LICENSE) that can be used in embedded applications to securely connect IoT devices to [AWS IoT Core](http://docs.aws.amazon.com/iot/latest/developerguide/what-is-aws-iot.html). It includes a MQTT, JSON Parser, and AWS IoT Device Shadow library. It is distributed in source form and intended to be built into customer firmware along with application code, other libraries and RTOS. 

## Features

> The Device SDK simplifies access to the Pub/Sub functionality of the AWS IoT broker via MQTT and provides APIs to interact with Device Shadow. The SDK has been tested to work with the AWS IoT Core and an open source MQTT broker to ensure best interoperability of a device with the AWS IoT platform. The C-SDK contains the following libraries:

Reviewer: C-SDK simplifies...

### AWS IoT Device Shadow
The [AWS IoT Device Shadow](https://github.com/aws/device-shadow-for-aws-iot-embedded-sdk) enables you to store and retrieve the current state (the “shadow”) of every registered device. The device’s shadow is a persistent, virtual representation of your device that you can interact with from AWS IoT Core even if the device is offline. The device state captured as its “shadow” is itself a [JSON](https://www.json.org/) document. You can send commands over MQTT or HTTP to retrieve the latest known device state, or to change the state. Each device’s shadow is uniquely identified by the name of the corresponding “thing”, a representation of a specific device or logical entity on the AWS Cloud. See [Managing Devices with AWS IoT](https://docs.aws.amazon.com/iot/latest/developerguide/iot-thing-management.html) for more information. More details about shadows can be found in [AWS IoT documentation](https://docs.aws.amazon.com/iot/latest/developerguide/iot-device-shadows.html). 

## Metrics

> Within the MQTT Demo, Users have the ability to report Operating System, Hardware Platform and MQTT Client information to AWS IoT by sending a specially formatted string in the username field of the MQTT CONNECT packet.  

Reviewer: Within the MQTT demo, users have the ability to report operating system, hardware platform and MQTT client information to AWS IoT by sending a specially formatted string in the username field of the MQTT CONNECT packet.

### Format

The format of the username string with metrics is:
```
<Actual_Username>?SDK=<OS_Name>&Version=<OS_Version>&Platform=<Hardware_Platform>&MQTTLib=<MQTT_Library_name>@<MQTT_Library_version>
```
where

* **Actual_Username** is the actual username used for authentication (if a username/password is used for authentication).
* **OS_Name** is the Operating System the application is running on.
* **OS_Version** is the version number of the Operating System.
* **Hardware_Platform** is the Hardware Platform the application is running on.
* **MQTT_Library_name** is the MQTT Client library being used.
* **MQTT_Library_version** is the version of the MQTT Client library being used.

> 

Reviewer: How do customers disable metrics? Can we add a line?

  
If you are upgrading from v3.x of the C-SDK to the 202009.00 release, please follow the update guide (LINK REQUIRED). If you are using the C-SDK v4_beta_deprecated branch, note that we will continue to fix bugs but will not add new features to it. See the C-SDK v4_beta_deprecated branch [README](https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/v4_beta_deprecated/README.md) for additional details. 

## Overview

> The AWS IoT Device SDK for Embedded C (C-SDK) is a collection of C source files under the [MIT open source license](LICENSE) that can be used in embedded applications to securely connect IoT devices to [AWS IoT Core](http://docs.aws.amazon.com/iot/latest/developerguide/what-is-aws-iot.html). It includes MQTT, JSON Parser, and AWS IoT Device Shadow. It is distributed in source form and intended to be built into customer firmware along with application code, other libraries and RTOS. For additional information about porting the C-SDK onto additional platforms, see the [Porting Guide](https://github.com/aws/aws-iot-device-sdk-embedded-c/blob/master/PortingGuide.md/).

Reviewer 2: Will the porting guide be added later as an .md file? RIght now, we only have doxygen docs for porting, and they're in each individual repo

## Features

> The Device SDK simplifies access to the Pub/Sub functionality of the AWS IoT broker and an open source MQTT broker (Mosquitto) via MQTT and provide APIs to interact with Device Shadow. The SDK has been tested to work with the AWS IoT Core to ensure best interoperability of a device with the AWS IoT platform. C-SDK contains the following libraries:

Reviewer 2: Do we need to mention mosquitto specifically? Also some articles and other things:
```suggestion
The Device SDK simplifies access to the Pub/Sub functionality of the AWS IoT broker, or an open source MQTT broker (such as Mosquitto), via MQTT, and provides APIs to interact with Device Shadows. The SDK has been tested to work with the AWS IoT Core to ensure best interoperability of a device with the AWS IoT platform. The C-SDK contains the following libraries:
```

### MQTT 

> The Device SDK provides the ability to establish an MQTT connection with a broker over a customer-implemented transport layer, which can either be a secure channel like a TLS session (mutually authenticated or server-only authentication) or a non-secure channel like a plaintext TCP connection. This MQTT connection can be used for performing publish operations to MQTT topics, and subscribing to MQTT topics. The SDK provides a mechanism to register customer-defined callbacks for receiving incoming PUBLISH, acknowledgement and keep-alive response events from the broker.  The [coreMQTT](https://github.com/FreeRTOS/coreMQTT) Client library has been refactored for memory optimization and implements the complete [MQTT 3.1.1](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/mqtt-v3.1.1.html) standard.  It has no dependencies on any additional libraries other than the standard C library, a customer-implemented network transport interface, and *optionally* a customer-implemented platform time function .  The refactored design embraces different use-cases, ranging from resource-constrained platforms using only QoS 0 MQTT PUBLISH messages to resource-rich platforms using QoS 2 MQTT PUBLISH over TLS connections.  

Reviewer 2: I think some plurals would be useful, as well as Oxford comma - unless I'm misinterpreting what you mean?
```suggestion
The Device SDK provides the ability to establish an MQTT connection with a broker over a customer-implemented transport layer, which can either be a secure channel like a TLS session (mutually authenticated or server-only authentication) or a non-secure channel like a plaintext TCP connection. This MQTT connection can be used for performing publish operations to MQTT topics, and subscribing to MQTT topics. The SDK provides a mechanism to register customer-defined callbacks for receiving incoming PUBLISH packets, acknowledgments, and keep-alive response events from the broker.  The [coreMQTT](https://github.com/FreeRTOS/coreMQTT) Client library has been refactored for memory optimization and implements the complete [MQTT 3.1.1](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/mqtt-v3.1.1.html) standard.  It has no dependencies on any additional libraries other than the standard C library, a customer-implemented network transport interface, and *optionally* a customer-implemented platform time function .  The refactored design embraces different use-cases, ranging from resource-constrained platforms using only QoS 0 MQTT PUBLISH messages to resource-rich platforms using QoS 2 MQTT PUBLISH over TLS connections.  
```

> The Device SDK provides the ability to establish an MQTT connection with a broker over a customer-implemented transport layer, which can either be a secure channel like a TLS session (mutually authenticated or server-only authentication) or a non-secure channel like a plaintext TCP connection. This MQTT connection can be used for performing publish operations to MQTT topics, and subscribing to MQTT topics. The SDK provides a mechanism to register customer-defined callbacks for receiving incoming PUBLISH, acknowledgement and keep-alive response events from the broker.  The [coreMQTT](https://github.com/FreeRTOS/coreMQTT) Client library has been refactored for memory optimization and implements the complete [MQTT 3.1.1](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/mqtt-v3.1.1.html) standard.  It has no dependencies on any additional libraries other than the standard C library, a customer-implemented network transport interface, and *optionally* a customer-implemented platform time function .  The refactored design embraces different use-cases, ranging from resource-constrained platforms using only QoS 0 MQTT PUBLISH messages to resource-rich platforms using QoS 2 MQTT PUBLISH over TLS connections.  

Reviewer 3: lets talk about "MQTT 3.1.1"

### AWS IoT Device Shadow

> The [AWS IoT Device Shadow](https://github.com/aws/device-shadow-for-aws-iot-embedded-sdk) enables you to store and retrieve the current state (the “shadow”) of every registered device. The device’s shadow is a persistent, virtual representation of your device that you can interact with in your applications even if the device is offline. The device state captured as its “shadow” is itself a [JSON](https://www.json.org/) document. You can send commands over MQTT or HTTP to retrieve the latest known device state, or to change the state. Each device’s shadow is uniquely identified by the name of the corresponding “thing”, a representation of a specific device or logical entity on the AWS Cloud. See [Managing Devices with AWS IoT](https://docs.aws.amazon.com/iot/latest/developerguide/iot-thing-management.html) for more information. More details about shadows can be found in [AWS IoT documentation](https://docs.aws.amazon.com/iot/latest/developerguide/iot-device-shadows.html). 

Reviewer 3: "The device’s shadow is a persistent, virtual representation of your device that you can interact with in your applications even if the device is offline."

which application?

* **OS_Name** is the Operating System the application is running on.
* **OS_Version** is the version number of the Operating System.
* **Hardware_Platform** is the Hardware Platform the application is running on.
* **MQTT_Library_name** is the MQTT Client library being used.
* **MQTT_Library_version** is the version of the MQTT Client library being used.

## Branches

### master branch
The [master](https://github.com/aws/aws-iot-device-sdk-embedded-C/tree/master) branch hosts the continuous development of the AWS IoT Embedded C SDK libraries. Please be aware that the development at the tip of the master branch is continuously in progress and may have bugs. Consider using the tagged releases of the AWS IoT Embedded C SDK for production ready software.

### v4_beta_deprecated branch
The [v4_beta_deprecated](https://github.com/aws/aws-iot-device-sdk-embedded-C/tree/v4_beta_deprecated) branch contains a beta version of the AWS IoT Embedded C SDK libraries, which is now deprecated. The libraries in this branch will not be released. However, critical bugs will be fixed and tested. No new features will be added to this branch.

## Releases

> All of the released versions of the AWS IoT Embedded C SDK libraries are available as git tags. For example, the latest v3 SDK version is available at [tag 3.1.0](https://github.com/aws/aws-iot-device-sdk-embedded-C/tree/v3.1.0).  

Reviewer 2: Looks like some whitespace was added?

# AWS IoT Device SDK C 202009.00

> **[Link to API documentation]**    (LINK REQUIRED)

Reviewer 4: We need to makes sure we incorporate some changes suggested by Pavan.

> We have released version 202009.00 of the AWS IoT SDK for Embedded-C (C-SDK). This release contains refactored libraries, and is a new design from the C-SDK v3.x. This release includes refactored MQTT and AWS IoT Device Shadow libraries for memory usage and modularity, and includes dependent libraries via GitHub submoduling. These libraries have gone through code quality checks including for [GNU Complexity](https://www.gnu.org/software/complexity/manual/complexity.html), [MISRA coding standard](https://www.misra.org.uk/MISRAHome/MISRAC2012/tabid/196/Default.aspx), [Coverity statical analysis](https://scan.coverity.com/), and [AWS CBMC automated reasoning tool](https://www.youtube.com/watch?v=YwQHAPRhQkI&feature=youtu.be&t=1721) to ensure memory safety, thread safety and functional correctness proof.   

Reviewer 5: > "This release includes refactored MQTT and AWS IoT Device Shadow libraries for memory usage and modularity, and includes dependent libraries via GitHub submoduling."

This sentence can be unclear to the reader. Can it be something like: "These libraries include MQTT, JSON parser and AWS IoT Device Shadow libraries that have been refactored for modularity and memory usage. This repository includes these libraries through Git submoduling. "

> We have released version 202009.00 of the AWS IoT SDK for Embedded-C (C-SDK). This release contains refactored libraries, and is a new design from the C-SDK v3.x. This release includes refactored MQTT and AWS IoT Device Shadow libraries for memory usage and modularity, and includes dependent libraries via GitHub submoduling. These libraries have gone through code quality checks including for [GNU Complexity](https://www.gnu.org/software/complexity/manual/complexity.html), [MISRA coding standard](https://www.misra.org.uk/MISRAHome/MISRAC2012/tabid/196/Default.aspx), [Coverity statical analysis](https://scan.coverity.com/), and [AWS CBMC automated reasoning tool](https://www.youtube.com/watch?v=YwQHAPRhQkI&feature=youtu.be&t=1721) to ensure memory safety, thread safety and functional correctness proof.   

Reviewer 5: ```suggestion
We have released version 202009.00 of the AWS IoT Device SDK for Embedded-C (C-SDK). This release contains refactored libraries, and is a new design from the C-SDK v3.x. This release includes refactored MQTT and AWS IoT Device Shadow libraries for memory usage and modularity, and includes dependent libraries via GitHub submoduling. These libraries have gone through code quality checks including for [GNU Complexity](https://www.gnu.org/software/complexity/manual/complexity.html), [MISRA coding standard](https://www.misra.org.uk/MISRAHome/MISRAC2012/tabid/196/Default.aspx), [Coverity statical analysis](https://scan.coverity.com/), and [AWS CBMC automated reasoning tool](https://www.youtube.com/watch?v=YwQHAPRhQkI&feature=youtu.be&t=1721) to ensure memory safety, thread safety and functional correctness proof.   
```

> If you are upgrading from v3.x of the C-SDK to the 202009.00 release, please follow the update guide (LINK REQUIRED). If you are using the C-SDK v4_beta_deprecated branch, note that we will continue to fix bugs but will not add new features to it. See the C-SDK v4_beta branch [README](https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/v4_beta_deprecated/README.md) for additional details. 

Reviewer 5: ```suggestion
If you are upgrading from v3.x of the C-SDK to the 202009.00 release, please follow the update guide (LINK REQUIRED). If you are using the C-SDK v4_beta_deprecated branch, note that we will continue to fix bugs but will not add new features to it. See the C-SDK v4_beta_deprecated branch [README](https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/v4_beta_deprecated/README.md) for additional details. 
```

## Overview

> The AWS IoT Device SDK for Embedded C (C-SDK) is a collection of C source files under the [MIT open source license](LICENSE) that can be used in embedded applications to securely connect IoT devices to [AWS IoT Core](http://docs.aws.amazon.com/iot/latest/developerguide/what-is-aws-iot.html). It includes transport clients MQTT, TLS implementations, and examples for their use. It also supports AWS IoT specific features such as Device Shadow. It is distributed in source form and intended to be built into customer firmware along with application code, other libraries and RTOS. For additional information about porting the C-SDK onto additional platforms, see the [Porting Guide](https://github.com/aws/aws-iot-device-sdk-embedded-c/blob/master/PortingGuide.md/).

Reviewer 5: This paragraph is missing mention of the JSON parser library.

Also, it uses "transport clients MQTT" which is incorrect as MQTT is not a transport protocol but an application layer protocol...

Author: The JSON specifically was left off by Tanmoy.

## Features

The Device SDK simplifies access to the Pub/Sub functionality of the AWS IoT broker via MQTT and provide APIs to interact with Device Shadow. The SDK has been tested to work with the AWS IoT Core to ensure best interoperability of a device with the AWS IoT platform. C-SDK contains the following libraries:

### MQTT 

> The Device SDK provides functionality to create and maintain a mutually authenticated TLS connection over which it runs MQTT. This connection is used for any further publish operations and allow for subscribing to MQTT topics which will call a configurable callback function when these topics are received.  The [coreMQTT](https://github.com/FreeRTOS/coreMQTT) Client library has been refactored for memory optimization and implements the complete [MQTT 3.1.1](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/mqtt-v3.1.1.html) standard.  It has no dependencies on any additional libraries other than the standard C library and a customer-implemented network transport interface.  The refactored design embraces different use-cases, ranging from resource-constrained platforms using only QoS 0 MQTT PUBLISH messages to resource-rich platforms using QoS 2 MQTT PUBLISH over TLS connections.  

Reviewer 5: >It has no dependencies on any additional libraries other than the standard C library and a customer-implemented network transport interface

It also has dependency on a platform-specific "get time" function as can be seen [here](https://github.com/FreeRTOS/coreMQTT/blob/master/source/include/core_mqtt.h#L189)

> The Device SDK provides functionality to create and maintain a mutually authenticated TLS connection over which it runs MQTT. This connection is used for any further publish operations and allow for subscribing to MQTT topics which will call a configurable callback function when these topics are received.  The [coreMQTT](https://github.com/FreeRTOS/coreMQTT) Client library has been refactored for memory optimization and implements the complete [MQTT 3.1.1](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/mqtt-v3.1.1.html) standard.  It has no dependencies on any additional libraries other than the standard C library and a customer-implemented network transport interface.  The refactored design embraces different use-cases, ranging from resource-constrained platforms using only QoS 0 MQTT PUBLISH messages to resource-rich platforms using QoS 2 MQTT PUBLISH over TLS connections.  

Reviewer 5: >This connection is used for any further publish operations and allow for subscribing to MQTT topics which will call a configurable callback function when these topics are received

This sentence seems incorrectly formed.. It can be modified to something like
"_This MQTT connection can be used for performing publish operations to MQTT topics, and subscribing to MQTT topics. The SDK provides a mechanism to register customer-defined callbacks for receiving incoming PUBLISH, acknowledgement and keep-alive response events from the broker._"

> The Device SDK provides functionality to create and maintain a mutually authenticated TLS connection over which it runs MQTT. This connection is used for any further publish operations and allow for subscribing to MQTT topics which will call a configurable callback function when these topics are received.  The [coreMQTT](https://github.com/FreeRTOS/coreMQTT) Client library has been refactored for memory optimization and implements the complete [MQTT 3.1.1](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/mqtt-v3.1.1.html) standard.  It has no dependencies on any additional libraries other than the standard C library and a customer-implemented network transport interface.  The refactored design embraces different use-cases, ranging from resource-constrained platforms using only QoS 0 MQTT PUBLISH messages to resource-rich platforms using QoS 2 MQTT PUBLISH over TLS connections.  

Reviewer 5: > The Device SDK provides functionality to create and maintain a mutually authenticated TLS connection over which it runs MQTT.

This statement can be misleading to suggest that the SDK provides functionality for establishing TLS session, which the library does not provide at all.. It can modified to something like:
"_The Device SDK provides the ability to establish an MQTT connection with a broker over a customer-implemented transport layer, which can either be a secure channel like a TLS session (mutually authenticated or server-only authentication) or a non-secure channel like a plaintext TCP connection."

> 

Reviewer 5: Should there be a section for the JSON parser?

**[API documentation](https://docs.aws.amazon.com/freertos/latest/lib-ref/embedded-csdk/202009.00/lib-ref/index.html)**    

> We have released version 202009.00 of the AWS IoT Device SDK for Embedded-C (C-SDK). This release contains refactored libraries, and is a new design from the C-SDK v3.x. This release includes refactored MQTT and AWS IoT Device Shadow libraries for optimized memory usage and modularity, and includes dependent libraries via GitHub submoduling. These libraries have gone through code quality checks including for [GNU Complexity](https://www.gnu.org/software/complexity/manual/complexity.html), [MISRA coding standard](https://www.misra.org.uk/MISRAHome/MISRAC2012/tabid/196/Default.aspx), [Coverity statical analysis](https://scan.coverity.com/), and [AWS CBMC automated reasoning tool](https://www.youtube.com/watch?v=YwQHAPRhQkI&feature=youtu.be&t=1721) to validate memory safety and functional correctness proof.   

Reviewer: I dont think this adds any value. Delete it? "This release contains refactored libraries, and is a new design from the C-SDK v3.x."

All of the released versions of the AWS IoT Embedded C SDK libraries are available as git tags. For example, the latest v3 SDK version is available at [tag 3.1.0](https://github.com/aws/aws-iot-device-sdk-embedded-C/tree/v3.1.0).

## Versioning

C-SDK releases will now follow a date based versioning scheme with the format YYYYMM.NN, where: 

* Y represents the year.
* M represents the month.
* N represents the release order within the designated month (00 being the first release).
* A "major" denotation indicates the addition of new features or significant updates to multiple libraries. 

For example, a second release in June 2021 would be 202106.01. Although the SDK releases have moved to date-based versioning, each library within the SDK will still retain semantic versioning. In semantic versioning, the version number itself (X.Y.Z) indicates whether the release is a major, minor, or point release. You can use the semantic version of a library to assess the scope and impact of a new release on your application. 

## Cloning
This repo uses [Git Submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules) to bring in dependent components.

> Note: If you download the ZIP file provided by GitHub UI, you will not get the contents of the submodules. (The ZIP file is also not a valid git repository)

Reviewer 5: We will be generating a ZIP containing submodule files and uploading to GitHub when making the release of this repository...

Reviewer 2: This note will still be necessary. GitHub will provide its own .zip of the source code in the releases page, and in 'Code'->'Download ZIP' on the main page, regardless of whatever assets we upload.

Reviewer 5: Alternatively, this note can point to the ZIP generated by us to take the advantage of submodules included in the ZIP file

Note: If you download the ZIP file provided by GitHub UI, you will not get the contents of the submodules. (The ZIP file is also not a valid git repository)  If you download from the [Tags](https://github.com/aws/aws-iot-device-sdk-embedded-C/tags) or [Releases](https://github.com/aws/aws-iot-device-sdk-embedded-C/releases), you will get the entire repository (including the submodules)

To clone using HTTPS:
```
git clone https://github.com/aws/aws-iot-device-sdk-embedded-C.git --recurse-submodules
```
Using SSH:
```
git clone git@github.com:aws/aws-iot-device-sdk-embedded-C.git --recurse-submodules
```

If you have downloaded the repo without using the `--recurse-submodules` argument, you need to run:
```
git submodule update --init --recursive
```

## Building and Running Demos

  
The libraries in this SDK are not dependent on any operating system. However, the demos for the libraries in this SDK are built and tested on a Linux platform. This SDK builds with [CMake](https://cmake.org/), a cross-platform build tool.

### Prerequisites

- CMake 3.13.0 or later 

> - C90 compiler

Reviewer 2: I think we should move line 105 (about `stdint.h`) to right below this. Needing `stdint.h` isn't a Linux-only caveat, so we should list it as a prerequisite

All of the released versions of the AWS IoT Embedded C SDK libraries are available as git tags. For example, the last release of the v3 SDK version is available at [tag 3.0.1](https://github.com/aws/aws-iot-device-sdk-embedded-C/tree/v3.0.1).

## Versioning

C-SDK releases will now follow a date based versioning scheme with the format YYYYMM.NN_major, where: 

* Y represents the year.
* M represents the month.
* N represents the release order within the designated month (00 being the first release).
* A "major" denotation indicates the addition of new features or significant updates to multiple libraries. 

For example, a second release in June 2021 would be 202106.01. Although the SDK releases have moved to date-based versioning, each library within the SDK will still retain semantic versioning. In semantic versioning, the version number itself (X.Y.Z) indicates whether the release is a major, minor, or point release. You can use the semantic version of a library to assess the scope and impact of a new release on your application. 

## Cloning
This repo uses [Git Submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules) to bring in dependent components.

> Note: If you download the ZIP file provided by GitHub UI, you will not get the contents of the submodules. (The ZIP file is also not a valid git repository)  If you download from the [Tags](https://github.com/aws/aws-iot-device-sdk-embedded-C/tags) or [Releases](https://github.com/aws/aws-iot-device-sdk-embedded-C/releases), you will get the entire repository (including the submodules)

Reviewer 2: I think the period goes after the parentheses. Also, should we specify which zip includes the entire repository? When there's a GitHub release, GitHub also has a "Source code (zip)" in addition to whatever files we upload as assets, and I think that one is identical to the one you get from the GitHub UI (e.g. [FreeRTOS](https://github.com/FreeRTOS/FreeRTOS/releases))
```suggestion
Note: If you download the ZIP file provided by GitHub UI, you will not get the contents of the submodules (The ZIP file is also not a valid git repository).  If you download from the [Tags](https://github.com/aws/aws-iot-device-sdk-embedded-C/tags) or [Releases](https://github.com/aws/aws-iot-device-sdk-embedded-C/releases), you will get the entire repository (including the submodules)
```

Reviewer 5: I agree that we should clearly mention that the "Source code (zip)" and "Source code (tar.gz)" assets that are auto-uploaded by GitHub do not contain the submodule code. 
We should also mention the name of the assert that we plan to upload, which will be named "aws-iot-device-sdk-embedded-c-202009.00.zip".
@leegeth, can you confirm the name of the new asset that we will upload?

Reviewer 6: Thanks Archit. That is a good point. The asset name is **aws-iot-device-sdk-embedded-C-202009.00.zip**

```shell
cd <CSDK_ROOT>
python3 docs/doxygen/generate_docs.py --root .
python3 tools/doxygen/generate_docs.py --root .
```

> The generated documentation landing page is located at **docs/doxygen/output/html/index.html**.

Reviewer 2: Was this line supposed to have been deleted? It seems like it's part of the documentation section above

> This library is licensed under the [MIT License](LICENSE).

Reviewer: C-SDK libraries are licensed....
